// @sapphire/discord.js-utilities/dist/lib/utilities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.canJoinVoiceChannel = exports.canRemoveAllReactions = exports.canReact = exports.canSendAttachments = exports.canSendEmbeds = exports.canSendMessages = exports.canReadMessages = void 0;
const utilities_1 = require("@sapphire/utilities");
const discord_js_1 = require("discord.js");
const type_guards_1 = require("./type-guards");
const canReadMessagesPermissions = new discord_js_1.PermissionsBitField([discord_js_1.PermissionFlagsBits.ViewChannel]);
/**
 * Determines whether or not we can read messages in a given channel.
 * @param channel The channel to test the permissions from.
 * @returns Whether or not we can read messages in the specified channel.
 */
function canReadMessages(channel) {
    if ((0, utilities_1.isNullish)(channel))
        return false;
    if ((0, type_guards_1.isDMChannel)(channel))
        return true;
    return canDoUtility(channel, canReadMessagesPermissions);
}
exports.canReadMessages = canReadMessages;
const canSendMessagesPermissions = new discord_js_1.PermissionsBitField([canReadMessagesPermissions, discord_js_1.PermissionFlagsBits.SendMessages]);
/**
 * Determines whether or not we can send messages in a given channel.
 * @param channel The channel to test the permissions from.
 * @returns Whether or not we can send messages in the specified channel.
 */
function canSendMessages(channel) {
    if ((0, utilities_1.isNullish)(channel))
        return false;
    if ((0, type_guards_1.isDMChannel)(channel))
        return true;
    if (channel.isThread() && !channel.sendable)
        return false;
    return canDoUtility(channel, canSendMessagesPermissions);
}
exports.canSendMessages = canSendMessages;
const canSendEmbedsPermissions = new discord_js_1.PermissionsBitField([canSendMessagesPermissions, discord_js_1.PermissionFlagsBits.EmbedLinks]);
/**
 * Determines whether or not we can send embeds in a given channel.
 * @param channel The channel to test the permissions from.
 * @returns Whether or not we can send embeds in the specified channel.
 */
function canSendEmbeds(channel) {
    if ((0, utilities_1.isNullish)(channel))
        return false;
    if ((0, type_guards_1.isDMChannel)(channel))
        return true;
    if (channel.isThread() && !channel.sendable)
        return false;
    return canDoUtility(channel, canSendEmbedsPermissions);
}
exports.canSendEmbeds = canSendEmbeds;
const canSendAttachmentsPermissions = new discord_js_1.PermissionsBitField([canSendMessagesPermissions, discord_js_1.PermissionFlagsBits.AttachFiles]);
/**
 * Determines whether or not we can send attachments in a given channel.
 * @param channel The channel to test the permissions from.
 * @returns Whether or not we can send attachments in the specified channel.
 */
function canSendAttachments(channel) {
    if ((0, utilities_1.isNullish)(channel))
        return false;
    if ((0, type_guards_1.isDMChannel)(channel))
        return true;
    if (channel.isThread() && !channel.sendable)
        return false;
    return canDoUtility(channel, canSendAttachmentsPermissions);
}
exports.canSendAttachments = canSendAttachments;
const canReactPermissions = new discord_js_1.PermissionsBitField([
    canSendMessagesPermissions,
    discord_js_1.PermissionFlagsBits.ReadMessageHistory,
    discord_js_1.PermissionFlagsBits.AddReactions
]);
/**
 * Determines whether or not we can send react to messages in a given channel.
 * @param channel The channel to test the permissions from.
 * @returns Whether or not we can react to messages in the specified channel.
 */
function canReact(channel) {
    if ((0, utilities_1.isNullish)(channel))
        return false;
    if ((0, type_guards_1.isDMChannel)(channel))
        return true;
    if (channel.isThread() && channel.archived)
        return false;
    return canDoUtility(channel, canReactPermissions);
}
exports.canReact = canReact;
const canRemoveAllReactionsPermissions = new discord_js_1.PermissionsBitField([
    canReadMessagesPermissions,
    discord_js_1.PermissionFlagsBits.ReadMessageHistory,
    discord_js_1.PermissionFlagsBits.ManageMessages
]);
/**
 * Determines whether or not we can remove reactions from messages in a given channel.
 * @param channel The channel to test the permissions from.
 * @returns Whether or not we can remove reactions from messages in the specified channel.
 */
function canRemoveAllReactions(channel) {
    if ((0, utilities_1.isNullish)(channel))
        return false;
    if ((0, type_guards_1.isDMChannel)(channel))
        return false;
    return canDoUtility(channel, canRemoveAllReactionsPermissions);
}
exports.canRemoveAllReactions = canRemoveAllReactions;
const canJoinVoiceChannelPermissions = new discord_js_1.PermissionsBitField([discord_js_1.PermissionFlagsBits.Connect]);
/**
 * Determines whether the client can join the given voice based channel.
 * @param channel The channel to test the permissions from.
 * @returns Whether or not the client can join the specified channel.
 */
function canJoinVoiceChannel(channel) {
    if ((0, utilities_1.isNullish)(channel))
        return false;
    if (!(0, type_guards_1.isVoiceBasedChannel)(channel))
        return false;
    if (channel.userLimit >= channel.members.size)
        return false;
    return canDoUtility(channel, canJoinVoiceChannelPermissions);
}
exports.canJoinVoiceChannel = canJoinVoiceChannel;
function canDoUtility(channel, permissionsToPass) {
    if (!(0, type_guards_1.isGuildBasedChannel)(channel)) {
        return true;
    }
    const { me } = channel.guild.members;
    if (!me)
        return false;
    const permissionsFor = channel.permissionsFor(me);
    if (!permissionsFor)
        return false;
    return permissionsFor.has(permissionsToPass);
}
